app/controllers/media_objects_controller.rb

```diff
@@ -54,6 +55,47 @@ def confirm_remove
     raise CanCan::AccessDenied unless Array(params[:id]).any? { |id| current_ability.can? :destroy, MediaObject.find(id) }
   end
 
+  def intercom_collections
+    reload = params['reload'] == 'true'
+    collections = session[:intercom_collections]
+    if reload || collections.blank?
+      intercom = Avalon::Intercom.new(user_key)
+      collections = intercom.user_collections
+      session[:intercom_collections] = collections
+    end
+    collections.each do |c|
+      c['default'] = c['id'] == session[:intercom_collection]
+    end
+    respond_to do |format|
+      format.json do
+        render json: collections.to_json
+      end
+    end
+  end
+
+  def intercom_push
+    if can? :intercom_push, @media_object
+      intercom = Avalon::Intercom.new(user_key)
+      collections = intercom.user_collections
+      session[:intercom_collections] = collections
+      if collections.collect { |c| c['id'] }.include? params[:collection_id]
```

Comment:
What about moving this checking into `Avalon::Intercom#push_media_object` and have that method return these different error messages?  That way the check can happen for any push using that method instead of just ones coming from the controller.  It might also make code climate happy.